lib/realtime/train_vehicles_pub_sub.ex

```diff
@@ -0,0 +1,177 @@
+defmodule Realtime.TrainVehiclesPubSub do
+  @moduledoc """
+  Saves train vehicles by route ID for sending to the client via the TrainVehiclesChannel.
+  """
+  use GenServer
+
+  alias Gtfs.Route
+  alias TrainVehicles.TrainVehicle
+
+  @type t :: %__MODULE__{
+          train_vehicles_by_route_id: train_vehicles_by_route_id()
+        }
+
+  @type train_vehicles_by_route_id :: %{Route.id() => [TrainVehicle.t()]}
+
+  @type broadcast_message :: {:new_train_vehicles, [TrainVehicle.t()]}
+
+  defstruct train_vehicles_by_route_id: %{}
+
+  # Client
+
+  @spec default_name() :: GenServer.name()
+  def default_name(), do: Realtime.TrainVehiclesPubSub
+
+  @spec start_link() :: GenServer.on_start()
+  @spec start_link(Keyword.t()) :: GenServer.on_start()
+  def start_link(opts \\ []) do
+    name = Keyword.get(opts, :name, default_name())
+
+    GenServer.start_link(
+      __MODULE__,
+      opts,
+      name: name
+    )
+  end
+
+  @spec subscribe(Route.id()) :: [TrainVehicle.t()]
+  @spec subscribe(Route.id(), GenServer.server()) :: [TrainVehicle.t()]
+  def subscribe(route_id, server \\ default_name()) do
+    {registry_key, train_vehicles} = GenServer.call(server, {:subscribe, route_id})
+    Registry.register(Realtime.Registry, registry_key, route_id)
+    train_vehicles
+  end
+
+  # Server
+
+  @impl GenServer
+  def init(opts) do
+    subscribe_fn = Keyword.get(opts, :subscribe_fn, &Phoenix.PubSub.subscribe/2)
+    subscribe_fn.(TrainVehicles.PubSub, "train_vehicles")
+
+    {:ok, %__MODULE__{}}
+  end
+
+  @impl GenServer
+  def handle_call(
+        {:subscribe, route_id},
+        _from,
+        %__MODULE__{train_vehicles_by_route_id: train_vehicles_by_route_id} = state
+      ) do
+    registry_key = self()
+
+    {:reply, {registry_key, train_vehicles_for_route_id(train_vehicles_by_route_id, route_id)},
+     state}
+  end
+
+  @impl GenServer
+  def handle_info({:reset, train_vehicles}, state) do
+    train_vehicles_by_route_id =
+      train_vehicles
+      |> Enum.group_by(&route_id_merging_green_lines/1)
+
+    new_state = Map.put(state, :train_vehicles_by_route_id, train_vehicles_by_route_id)
+
+    _ = broadcast(new_state)
+
+    {:noreply, new_state}
+  end
+
+  def handle_info(
+        {:add, new_train_vehicles},
```

Comment:
This module has two major responsibilities:
* broadcasting new data to channels.
* updating the list of known vehicles as streaming data trickles in

Both are pretty complex tasks. Could we pull the "updating with new stream data" into a separate module (that's just functions, not a gen server). It'd have an interface like `@spec update_data(train_vehicles_by_route_id(), event(), new_data()) :: train_vehicles_by_route_id())`

Reply:
Good idea, I refactored out a `TrainVehiclesStore`.